specs/tree-view/tree-view.md

```diff
@@ -0,0 +1,186 @@
+
+### Use Cases
+
+Primarily used to present a hierarchical list such as a file system. A tree view can also be served as a flat list.
+  
+### Features
+
+- **Selection:** A tree view can support either single select or multi-select behavior for nodes. **I would propose that for the initial spec we start with supporting single select only.*
+
+
+- **Drag & Drop support:** Often, as is in the case of a "file system" example, drag and drop support is desired. This does present [challenges](#drag-&-drop).
+
+### Risks and Challenges
+
+#### Nesting:
+
+Infinite nesting combined with the current visual design presents a styling challenge. Consider a scenario where we have the following tree:
+```html
+<fast-tree-view>
+    <fast-tree-item>Root node one</fast-tree-item>
+    <fast-tree-item>Root node two</fast-tree-item>
+    <fast-tree-item>
+        Parent Node && Root node
+        <fast-tree-item slot="child-nodes">Nested node one</fast-tree-item slot="child-nodes">
+        <fast-tree-item slot="child-nodes">Nested node two</fast-tree-item>
+    </fast-tree-item>
+    <fast-tree-item>Root node three</fast-tree-item>
+</fast-tree-view>
+```
+In this scenario, there is no root node with an expand/collapse glyph. The third item in the tree view is a parent node, which *would* have an expand/collapse glyph to the right. To ensure that the content of the parent node align with the rest of the root nodes, the entire tree will need to visually shift to ensure alignmnet of the tree content. This presents an interesting challenge for how we will communicate these changes to the parent to ensure the styling is appropriately applied. Ultimately, the parent just needs to know that *one* of the items below has child nodes.
+
+#### Drag & Drop
+
+Drag & Drop support poses a unique set of challenges, especially if our goal is to provide out of the box support via the [HTML Drag and Drop API](https://developer.mozilla.org/en-US/docs/Web/API/HTML_Drag_and_Drop_API). Alternatively, those using our components with React, Vue, Angular, etc may want to use libraries built to work with their specific framework flavor. 
+
+**Considering the challenge that drag & drop would present for any given component, I think it's worth holding on that feature, and taking it up seperately.*
+
+### Prior Art/Examples
+- [FAST-DNA (React)](https://explore.fast.design/components/tree-view)
+- [Material UI](https://material-ui.com/components/tree-view/)
+- [Lightning Design](https://www.lightningdesignsystem.com/components/trees/)
+- [Ant Design](https://ant.design/components/tree/)
+- [Atlassian](https://atlaskit.atlassian.com/packages/core/tree/example/static-tree)
+- [Windows (UWP)](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/tree-view)
+
+---
+
+## Design
+
+### API
+
+*The key elements of the component's public API surface:*
+
+**Tree View**
+
+- extends HTML Element props
+- custom CSS property for `depth` to support the infinite nesting scenario
```

Comment:
Is the idea here that `depth` would be incremented at each level and then used to calculate some indentation in styles?

Reply:
Yes, that's the overall concept here. Thoguhts/concerns?

Reply:
The only trick would be the need to read the depth property before calculating the new depth, which would require a getComputedStyles call...which would force layout, etc. 

Another approach might be to use the connectedCallback of `tree-view-item` to have it look at it's parent element and check a property that specifies depth. Then it could calc and set the same property on itself. That property could then be used to apply an inline style...or just set a css property as well. We can experiment and see what ends up being better.

Reply:
Sure - I don't think we'll need the getComputedStyles here as I think there's a better way, but we will need to experiment.

Reply:
Wow. I can't believe css vars can't self-reference, I thought I'd seen this in the wild and thought for sure this would work:

`style="--depth: calc(var(--depth) + 1)`.

Here is a less-optimal solution: https://codesandbox.io/s/stoic-brook-73o1t?fontsize=14&hidenavigation=1&theme=dark

It uses two elements and two depth variables. It works but isn't ideal. Might be better then manually tracking a property in JS though.